### PR TITLE
Keycloak to serverless

### LINKED ISSUE #53: Lambdas for keycloak users

### PR DESCRIPTION
**Pull request by Iiro127**

### CODE REVIEW

File: src/functions/Questionnaires/find-users/handler.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      return {
+        statusCode: 404,
+        body: JSON.stringify({ error: 'User not found' }),
+      };
+    }
+  } catch (error) {
+    console.error('DynamoDB error finding software:', error);
+    return {
+      statusCode: 500,
+      body: JSON.stringify({ error: 'Failed to retrieve software.', details: error.message }),
```
Comment (McDuuck) [inline src/functions/Questionnaires/find-users/handler.ts]: double quotes

File: src/functions/Questionnaires/find-users/handler.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+const 
+
+export const findUsersHandler: APIGatewayProxyHandler = async (event: APIGatewayProxyEvent) => {
+  const { id } = event.pathParameters || {};
+
+  if (!id) {
+    return {
+      statusCode: 400,
+      body: JSON.stringify({ error: 'Missing or invalid path parameter: id' }),
```
Comment (McDuuck) [inline src/functions/Questionnaires/find-users/handler.ts]: double quotes

File: src/functions/keycloak/list-users/index.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { handlerPath } from "src/libs/handler-resolver";
+
+export default {
+    handler: `${handlerPath(__dirname)}/handler.main`,
+    events: [
+        {
+            httpApi: {
+                method: 'get',
```
Comment (dalju005-xamk) [inline src/functions/keycloak/list-users/index.ts]: should it be GET

File: src/functions/keycloak/list-users/handler.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                firstName: user.firstName,
+                lastName: user.lastName,
+                email: user.email,
+                isActive: user.isActive,
+                severaGuid: user.severaGuid,
+                forecastId: user.forecastId
+            }
+    })
+    } catch (error) {
+        throw error()
```
Comment (dalju005-xamk) [inline src/functions/keycloak/list-users/handler.ts]: adding some message to error ?

**Review by dalju005-xamk — CHANGES_REQUESTED** [review]

File: src/functions/keycloak/list-users/index.ts (hunk 3 above)
Comment (Iiro127) [reply src/functions/keycloak/list-users/index.ts]: I don't think so

**Review by dalju005-xamk — APPROVED** [review]

File: src/functions/keycloak/find-user/handler.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ * Gets user from keycloak
+ *
+ * @param api
+ * @param id
+ * @returns user information by Id
+ */
+const findUser = async (api: KeycloakApiService, id: string): Promise<Response[]> => {
+  const user = await api.findUser(id);
+  if (!user) {
+    throw new Error('Cannot find user from Api');
```
Comment (McDuuck) [inline src/functions/keycloak/find-user/handler.ts]: double quotes

File: src/apis/keycloak-api-service.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                throw new Error(`Failed to find user with id: ${id}`);
+            }
+
+            const userData = await response.json();
+
+            return userData;
+        },
+    }
+}
+
```
Comment (McDuuck) [inline src/apis/keycloak-api-service.ts:82]: Create TSDocs like others have, for example : /**
         * Find user from keycloak
         * 
         * @returns user by Id
         */

**Review by McDuuck — CHANGES_REQUESTED** [review]

File: src/functions/keycloak/find-user/handler.ts (hunk 5 above)
Comment (dalju005-xamk) [reply src/functions/keycloak/find-user/handler.ts]: "Checked"

**Review by McDuuck — APPROVED** [review]

File: serverless.ts [hunk 7]
```diff
@@ -26,6 +26,10 @@ import findSoftwareHandler from "@/functions/software-registry/find-software";
 import listSoftwareHandler from "@/functions/software-registry/list-software";
 import updateSoftwareHandler from "@/functions/software-registry/update-software";
 import deleteSoftwareHandler from "@/functions/software-registry/delete-software";
+import listUsersHandler from "@/functions/keycloak/list-users"
+import findUserHandler from "src/functions/keycloak/find-user";
+import testFunctions from "src/functions/test"
+
```
Comment (Jdallos) [inline serverless.ts]: No need for double line break

Comment (Jdallos) [inline serverless.ts]:
```suggestion
```

File: serverless.ts [hunk 8]
```diff
@@ -140,7 +144,10 @@ const serverlessConfiguration: AWS = {
     findSoftwareHandler,
     listSoftwareHandler,
     updateSoftwareHandler,
-    deleteSoftwareHandler
+    deleteSoftwareHandler,
+    listUsersHandler,
+    findUserHandler,
+    testFunctions
```
Comment (Jdallos) [inline serverless.ts]: Should test functions be a part of this PR? If you think they should then they should have a more descriptive name otherwise be removed

File: src/apis/keycloak-api-service.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import type { User } from "./schemas/keycloak/user";
+
+export interface KeycloakApiService {
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts:6]: Add Docs for the interface

File: src/apis/keycloak-api-service.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    getUsers: () => Promise<User[]>,
+    findUser: (id: string) => Promise<User[]>
+}
+
+/**
+ * Creates KeycloakApiService
+ */
+export function CreateKeycloakApiService(): KeycloakApiService {
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Use arrow functions unless there is a specific reason not to.

File: src/apis/keycloak-api-service.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        async getUsers(): Promise<User[]> {
+            const response = await fetch(`${baseUrl}/admin/realms/${realm}/users`, {
+                method: 'GET',
+                headers: {
+                    'Authorization': `Bearer ${await getAccessToken()}`
+                }
+            })
+
+            if (!response.ok) {
+                console.log(`Failed to fetch users: ${response.status} - ${response.statusText}`);
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Maybe this should be a console.error?

File: src/apis/keycloak-api-service.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    const baseUrl: string = process.env.KEYCLOAK_BASE_URL
+    const realm: string = process.env.KEYCLOAK_REALM
+
+    return {
+        /**
+         * Gets all users from keycloak
+         * 
+         * @returns List of users
+         */
+        async getUsers(): Promise<User[]> {
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: It looks like your indentation is set to 4 space or similar- please set to 2 spaces and update all files with updated indentation

File: src/apis/keycloak-api-service.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            if (!response.ok) {
+                console.log(`Failed to fetch users: ${response.status} - ${response.statusText}`);
+            }
+
+            return response.json();
+        },
+
+        /**
+         * Find user from keycloak
+         *
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Missing @param Doc

File: src/apis/keycloak-api-service.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        async findUser(id: string): Promise<User[]> {
+            const response = await fetch(`${baseUrl}/admin/realms/${realm}/users/${id}`, {
+                method: 'GET',
+                headers: {
+                    'Authorization': `Bearer ${await getAccessToken()}`
+                } 
+            })
+
+            if (!response.ok) {
+                throw new Error(`Failed to find user with id: ${id}`);
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Be consistent across functions, either console.error or throw new Error but have the same behavior

File: src/apis/keycloak-api-service.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        },
+    }
+}
+
+/**
+ * Requests an access token from keycloak API
+ * 
+ * @returns access token as string
+ */
+async function getAccessToken(): Promise<string> {
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Use arrow function

File: src/apis/keycloak-api-service.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    const realm: string = process.env.KEYCLOAK_REALM
+    const url: string = `${process.env.KEYCLOAK_BASE_URL}/realms/${realm}/protocol/openid-connect/token`
+    const requestBody = new URLSearchParams ({
+        'client_id': process.env.KEYCLOAK_CLIENT_ID,
+        'client_secret': process.env.KEYCLOAK_CLIENT_SECRET,
+        'username': process.env.KEYCLOAK_ADMIN_USERNAME,
+        'password': process.env.KEYCLOAK_ADMIN_PASSWORD,
+        'grant_type': "client_credentials"
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Use double quotes

File: src/apis/keycloak-api-service.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    });
+
+    try {
+        const response = await fetch(url, {
+            method: 'POST',
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Double quotes

File: src/apis/keycloak-api-service.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+            headers: {
+                'Content-Type': 'application/x-www-form-urlencoded'
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Quotes

File: src/apis/keycloak-api-service.ts (hunk 15 above)
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]: Lets check with Antti if it is fine to call this function everytime we want to make a request to the API

File: src/apis/keycloak-api-service.ts [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Requests an access token from keycloak API
+ * 
+ * @returns access token as string
+ */
+async function getAccessToken(): Promise<string> {
+    const realm: string = process.env.KEYCLOAK_REALM
+    const url: string = `${process.env.KEYCLOAK_BASE_URL}/realms/${realm}/protocol/openid-connect/token`
+    const requestBody = new URLSearchParams ({
```
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts]:
```suggestion
    const requestBody = new URLSearchParams({
```

File: src/apis/schemas/keycloak/user.ts [hunk 20]
```diff
@@ -0,0 +1,9 @@
+export interface User {
```
Comment (Jdallos) [inline src/apis/schemas/keycloak/user.ts:4]: Add Docs

Comment (Jdallos) [reply src/apis/schemas/keycloak/user.ts:4]: Semi colons please

File: src/functions/Questionnaires/find-users/handler.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import { APIGatewayProxyEvent, APIGatewayProxyHandler } from "aws-lambda";
+import { DocumentClient } from "aws-sdk/clients/dynamodb";
+import { middyfy } from "src/libs/lambda";
+
+const
```
Comment (Jdallos) [inline src/functions/Questionnaires/find-users/handler.ts]:
```suggestion
```

Comment (Jdallos) [reply src/functions/Questionnaires/find-users/handler.ts]: Remove or fix

File: src/functions/Questionnaires/find-users/handler.ts [hunk 22]
```diff
@@ -0,0 +1,40 @@
+import { APIGatewayProxyEvent, APIGatewayProxyHandler } from "aws-lambda";
+import { DocumentClient } from "aws-sdk/clients/dynamodb";
+import { middyfy } from "src/libs/lambda";
+
+const 
+
+export const findUsersHandler: APIGatewayProxyHandler = async (event: APIGatewayProxyEvent) => {
```
Comment (Jdallos) [inline src/functions/Questionnaires/find-users/handler.ts]: Does this need to be exported?

File: src/functions/keycloak/find-user/handler.ts [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+  // Ensure queryStringParameters and id are present
+  if (!queryStringParameters || !queryStringParameters.id) {
+    return {
+      headers:{
+        "Content-Type": "application/json"
+      },
+      statusCode: 400,
+      body: JSON.stringify({ error: "Missing or invalid path parameter: id" })
+    };
+  }
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: How about having this within the try and handling the thrown error so that it is handled in the same catch you have below- you could pass the statusCode and error message?

File: src/functions/keycloak/find-user/handler.ts [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+  try {
+    const userById = await findUser(api, queryStringParameters.id);
+        if (!userById) {
+      return {
+        headers:{
+          "Content-Type": "application/json"
+        },
+        statusCode: 404,
+        body: JSON.stringify({ error: "User not found" })
+    };
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: This seems redundant as you have the check in the findUser function. I would suggest just calling the api.findUser from this handler function and then checking the userById, but again throwing so that the error is handled in the catch rather than having multiple error returns as you have currently.

File: src/functions/keycloak/find-user/handler.ts [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    return {
+      headers:{
+        "Content-Type": "application/json"
+      },
+      statusCode: 200,
+      body: JSON.stringify(userById)
+    };
+  } catch (error) {
+    console.log("query string parameters:", queryStringParameters);
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: console.error?

File: src/functions/keycloak/find-user/handler.ts [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    return {
+      headers:{
+        "Content-Type": "application/json"
+      },
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: Are you sure this is necessary?

File: src/functions/keycloak/list-users/handler.ts [hunk 27]
```diff
@@ -0,0 +1,67 @@
+import { middyfy } from "@libs/lambda";
+import type { APIGatewayProxyHandler } from "aws-lambda";
+import { CreateKeycloakApiService, type KeycloakApiService } from "src/apis/keycloak-api-service";
+
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]:
```suggestion
```

File: src/functions/keycloak/list-users/handler.ts [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+ * Response schema for lambda
+ */
+interface Response {
+    id: string
+    firstName: string
+    lastName: string
+    email: string
+    isActive: boolean
+    severaGuid: string
+    forecastId: number
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]: semi colons

File: src/functions/keycloak/list-users/handler.ts [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+                firstName: user.firstName,
+                lastName: user.lastName,
+                email: user.email,
+                isActive: user.isActive,
+                severaGuid: user.severaGuid,
+                forecastId: user.forecastId
+            }
+    })
+    } catch (error) {
+        throw error("listUsers(): Error when listing users")
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]:
```suggestion
        throw error("Error when listing users")
```

File: src/functions/keycloak/list-users/handler.ts [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+    }
+}
+
+/**
+ * Lambda for listing users
+ * 
+ * @param _event event
+ */
+const listUsersHandler: APIGatewayProxyHandler = async (_event) => {
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]:
```suggestion
const listUsersHandler: APIGatewayProxyHandler = async () => {
```

File: src/functions/keycloak/list-users/handler.ts [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+    })
+    } catch (error) {
+        throw error("listUsers(): Error when listing users")
+    }
+}
+
+/**
+ * Lambda for listing users
+ * 
+ * @param _event event
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]:
```suggestion
```

File: src/functions/keycloak/list-users/handler.ts [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+ */
+const listUsersHandler: APIGatewayProxyHandler = async (_event) => {
+    try {
+        const api = CreateKeycloakApiService();
+        const users = await listUsers(api)
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]: Similar to the find users I see no need for a separate listUsers function here as this function is small enough. Just call the api.listUsers from here and handle appropriately

File: src/functions/keycloak/list-users/handler.ts [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+
+        return {
+            statusCode: 200,
+            body: JSON.stringify({ users })
+        }
+    } catch (_error){
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]: Is there a reason you are not logging the error.message?

File: src/functions/keycloak/list-users/index.ts [hunk 34]
```diff
@@ -0,0 +1,16 @@
...
+                path: '/users',
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/index.ts]: Double quotes

File: src/functions/Questionnaires/find-users/handler.ts
Comment (Jdallos) [inline src/functions/Questionnaires/find-users/handler.ts:1]: I have not reviewed this file, shouldn't it be part of a separate PR? If so please remove.

File: src/functions/Questionnaires/find-users/index.ts
Comment (Jdallos) [inline src/functions/Questionnaires/find-users/index.ts:1]: I have not reviewed this file, shouldn't it be part of a separate PR? If so please remove.

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/apis/keycloak-api-service.ts
Comment (Jdallos) [inline src/apis/keycloak-api-service.ts:1]: Fix indentation in all files, it should be set to indent using spaces: 2 for all files, currently it is set to 4 in some files or tabs 2 in other files

File: src/functions/keycloak/list-users/handler.ts [hunk 35]
```diff
@@ -1,54 +1,53 @@
...
-                    severaGuid: user.severaGuid,
-                    forecastId: user.forecastId
-                };
-                return responseUser;
-            });
-        });
+  try {
+	  const api = CreateKeycloakApiService();
+		
+	  const users: Response[] = await api.getUsers().then((users) => {
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]: Typically we use async await OR .then.
And generally we use async await instead of .then.
So I'd recommed changing this to use just async await and not the .then.

This can be done by awaiting the getUsers() and assigning it to a named variable and then mapping as required,

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/apis/keycloak-api-service.ts (hunk 15 above)
Comment (Jdallos) [reply src/apis/keycloak-api-service.ts]: This is ok for now and will be changed in a future feature as we implement a keycloak client

File: src/functions/keycloak/find-user/handler.ts [hunk 36]
```diff
@@ -0,0 +1,64 @@
...
+ * @returns user information as string
+ */
+
+const findUserHandler: APIGatewayProxyHandler = async (event: APIGatewayProxyEvent) => {
+  let apiResponse;
+
+  try {
+    const { queryStringParameters } = event;
+
+    // Ensure queryStringParameters and id are present before making the API call
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: No need for this comment IMO

File: src/functions/keycloak/find-user/handler.ts [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+
+/** 
+ * Lambda for finding user
+ * 
+ * @param event event
+ * @returns user information as string
+ */
+
+const findUserHandler: APIGatewayProxyHandler = async (event: APIGatewayProxyEvent) => {
+  let apiResponse;
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: Is this actually needed,  can't you just return the appropriate objects without assigning them to this apiResponse (which is initialized as type any)- something we want to avoid.

File: src/functions/keycloak/find-user/handler.ts [hunk 38]
```diff
@@ -0,0 +1,64 @@
...
+
+  try {
+    const { queryStringParameters } = event;
+
+    // Ensure queryStringParameters and id are present before making the API call
+    if (!queryStringParameters || !queryStringParameters.id) {
+      throw new Error("Missing or invalid path parameter: id");
+    }
+    const api = CreateKeycloakApiService();
+    const userById = await findUser(api, queryStringParameters.id);
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/handler.ts]: I still don't see why you can't just call api.findUser(id) here.

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/functions/keycloak/list-users/handler.ts (hunk 35 above)
Comment (dalju005-xamk) [reply src/functions/keycloak/list-users/handler.ts]: Made changes to Iiros code. Now it has async await structure.

File: src/functions/keycloak/find-user/index.ts [hunk 39]
```diff
@@ -0,0 +1,16 @@
+import { handlerPath } from '@libs/handler-resolver';
+
+export default {
+  handler: `${handlerPath(__dirname)}/handler.main`,
+  events: [
+    {
+      httpApi: {
+        method: 'get',
+        path: '/users/{id}',
```
Comment (Jdallos) [inline src/functions/keycloak/find-user/index.ts]: Double quotes

File: src/functions/keycloak/list-users/handler.ts [hunk 40]
```diff
@@ -0,0 +1,56 @@
...
+      };
+      return responseUser;
+    });
+
+    return {
+      statusCode: 200,
+      body: JSON.stringify(mappedUsers),
+    };
+  } catch (error) {
+    console.log(error);
```
Comment (Jdallos) [inline src/functions/keycloak/list-users/handler.ts]: use console.error here

**Review by Jdallos — APPROVED** [review]
